### ISSUE TITLE
problem with get_pssm when preprocessing variants

### ISSUE THREAD

**Issue by rgayatri** [DeepRank/deeprank-core#429]
When check_pssm is ignored, I face another error with get_pssm 
```

Traceback (most recent call last):
  File "/home/gayatrir/miniconda3/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/gayatrir/miniconda3/lib/python3.10/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/mnt/home2/gayatrir/deeprank-core/deeprankcore/query.py", line 244, in _process_one_query
    graph = query.build(feature_modules)
  File "/mnt/home2/gayatrir/deeprank-core/deeprankcore/query.py", line 609, in build
    feature_module.add_features(self._pdb_path, graph, variant)
  File "/mnt/home2/gayatrir/deeprank-core/deeprankcore/features/conservation.py", line 29, in add_features
    pssm_row = residue.get_pssm()
  File "/mnt/home2/gayatrir/deeprank-core/deeprankcore/molstruct/residue.py", line 54, in get_pssm
    raise FileNotFoundError(f'No pssm file found for Chain {self._chain}.')
FileNotFoundError: No pssm file found for Chain pdb2g98 B
```

My variant of interest is in chain A, thus the pssm_path I provide is for chain A. 
Code to reproduce.

```
from deeprankcore.query import QueryCollection, SingleResidueVariantAtomicQuery
from deeprankcore.features import components, conservation, contact, surfacearea
from deeprankcore.utils.grid import GridSettings, MapMethod
from deeprankcore.domain.aminoacidlist import (arginine, cysteine)
from deeprankcore.domain import targetstorage as targets

queries.add(SingleResidueVariantAtomicQuery(
    pdb_path = "/home/gayatrir/DATA/pdb/pdb2g98.ent",
    chain_id = "A",
    residue_number = 14,
    insertion_code = None,
    wildtype_amino_acid = arginine,
    variant_amino_acid = cysteine,
    pssm_paths = {"A": "/mnt/csb/DeepRank-Mut-DATA/pssm_update/2g98.A.pdb.pssm"},
    targets={targets.BINARY: 1},
    radius= 10.0,
    distance_cutoff= 4.5,
    ))

feature_modules = [components, conservation, contact, surfacearea]

hdf5_paths = queries.process(
    "/home/gayatrir/coretrain.hdf5",
    feature_modules = feature_modules,
    grid_settings = GridSettings(
        points_counts = [20, 20, 20],
        sizes = [1.0, 1.0, 1.0]),
    grid_map_method = MapMethod.GAUSSIAN)


```

**Comment by DaniBodor** [comment]
OK, so ultimately the problem is that if the pdb file contains residues for a secondary chain that are within `radius` they will be nodes in the graph. All nodes for the graph need to have all the features as each other (inherent property of graphs). 

So, I see three ways to deal with this, but happy to hear other suggestions as well:
1. we do not include residues from the secondary chain in the graph
    - i think we need to change `build` method for this
1. we fill the pssm with 0s (or other 'dummy' value) for the residues of the secondary chain
    - i think this would be done in feature module and/or in molstruct/residue.py
1. we give a pssm for the secondary chain as well
    - for this the code should work already

Note, that I solved the problem for the `_check_pssm` function in #430, but it's kind of irrelevant after considering these issues.

**Comment by rgayatri** [comment]
Yes. However, not all variants may have nodes from the other chain within the radius. 
In DeepRank-Mut we dealt with this flexibly, where features of atoms/residues from another chain are included only when it is within the radius. 

I did try by including the other chains within radius, but the code tries to retrieve pssms for all chains in PDB. 
For instance, a PDB file has 4 chains- A, B, C, D, and there is a mutation in chain A with a radius where atoms from chain B are included, the code still looks for pssms for chains C and D.

- I know option 3 would work- but it doesn't seem optimal. 
- Option 2 can work too. So when the user specifies one chain, all other chains are filled with 0s? Is this what you mean?
- I think option 1 might not be helpful, as sometimes you need single chains and sometimes multiple depending on where the variant is.

**Comment by DaniBodor** [comment]
> Yes. However, not all variants may have nodes from the other chain within the radius.
> In DeepRank-Mut we dealt with this flexibly, where features of atoms/residues from another chain are included only when it is within the radius.
> I did try by including the other chains within radius, but the code tries to retrieve pssms for all chains in PDB.
> For instance, a PDB file has 4 chains- A, B, C, D, and there is a mutation in chain A with a radius where atoms from chain B are included, the code still looks for pssms for chains C and D.
> 
>     I know option 3 would work- but it doesn't seem optimal.
>     Option 2 can work too. So when the user specifies one chain, all other chains are filled with 0s? Is this what you mean?
>     I think option 1 might not be helpful, as sometimes you need single chains and sometimes multiple depending on where the variant is.
> 

That is indeed what I meant by option 2. If that is the preferred solution, I'll put it on the to do list.

**Comment by gcroci2** [comment]
In a meeting not about this issue I talked with Li about the error you were getting @rgayatri, and she told me that it should be up to the user to provide a PDB file that contains only the chain of interest. So maybe you can first discuss this with her and then let us know? 
Relevant to @DaniBodor as well

**Comment by rgayatri** [comment]
I think it is a bit too much to expect from a user to know exactly where the mutation is (for 100,000 cases), accordingly parse pdbs with single or multiple chains and then use deeprank-core to preprocess them. 
One often finds an easy way out- if this repository is only meant for PPIs I would rather leave out variants from this. 
In the end, its upto Li

**Comment by gcroci2** [comment]
Actually you do need to know where the mutation is (in which chain), in order to pass the right PSSM file, right? Otherwise you would pass all PSSMs. And also in the latter case, there should be no issue with the current code. @rgayatri 

We can always proceed with option 2 that @DaniBodor suggested.

@LilySnow Any comments about this discussion?

**Comment by rgayatri** [comment]
In DeepRank-Mut we pass directories of pdbs and pssms, and the code looks for single chain or multiple chain PSSMs wherever necessary. I can try running with all pssms for all chains.

**Comment by gcroci2** [comment]
> In DeepRank-Mut we pass directories of pdbs and pssms, and the code looks for single chain or multiple chain PSSMs wherever necessary. I can try running with all pssms for all chains.

So if you initialize a variant that happens to be in chain A, with a radius where atoms from chain B are included, the code should look only at PSSMs for chains A _and_ B, right? But not for other eventual other chains C and D.
And if you initialize a variant that happens to be in chain A, with a radius where no atoms from other chains are included, then you would need chain A only.
So we should look only at the PSSMs file of the chains present in the radius, that's what you do in DeepRank-Mut if I understood. @rgayatri

**Comment by rgayatri** [comment]
Yes that is correct @gcroci2

**Comment by DaniBodor** [comment]
I now have a simple solution for option 2 in #430. That can still be undone, obviously.

What is unclear to me is whether residues on other chains than where the mutation is could contribute. If yes, then in my opinion, we should require pssm files to be present. If for some cases/projects the pssm of the secondary chains are irrelevant, users can create pssm files with all 0s.
What we don't is to create a system where we allow users to input faulty data by accident or because they didn't consider it and be able to use the software without errors.

**Comment by LilySnow** [comment]
Sometimes the variant appears only in one chain and sometimes at the interface of multiple chains. So I think the user should remove unrelevant chains from the PDB and provide PSSMs for all relevant chains, right? I can also discuss with Gayatri tomorrow in person about it.

**Comment by LilySnow** [comment]
Gayatri and I chatted. There are some miscommunications. It seems that DeepRank-core warning/error code needs to be changed. 

The variant sometimes appears at the interface of multiple chains. For example, the variant is between chain A and chain B, and the PDB file contains chains A, B, C, and D. It is sufficient for the user to provide PSSM files for chain A and B. However, at this moment, DeepRank-core gives an error message if it does not find PSSM files for chain C and D. 

We think DeepRank-core should be able to process this situation.

**Comment by DaniBodor** [comment]
> Gayatri and I chatted. There are some miscommunications. It seems that DeepRank-core warning/error code needs to be changed.
> 
> The variant sometimes appears at the interface of multiple chains. For example, the variant is between chain A and chain B, and the PDB file contains chains A, B, C, and D. It is sufficient for the user to provide PSSM files for chain A and B. However, at this moment, DeepRank-core gives an error message if it does not find PSSM files for chain C and D.
> 
> We think DeepRank-core should be able to process this situation.

The problem you are raising here is actually issue #428, which I already solved it in PR #430 (still unmerged).

In this issue, @rgayatri raised a separate issue, which is if the variant appears at the interface of chain A and chain B but only a pssm file for chain A was provided, then there is also an error. [I gave three option to deal with this here](https://github.com/DeepRank/deeprank-core/issues/429#issuecomment-1557641636), and already implemented a solution for option 2 in #430. However, it appears we are gravitating more towards option 3, which is basically leaving it as is and requiring pssm files for all chains which are within the radius.

Please confirm how you would like me to proceed with this.
